dpdk/vr_dpdk_lcore.c

```diff
@@ -82,55 +82,66 @@ vr_dpdk_lcore_least_used_get(void)
     return least_used_id;
 }
 
-/* Add a RX queue to a lcore */
+/* Add a queue to a lcore */
 void
-dpdk_lcore_rx_queue_add(unsigned lcore_id, struct vr_dpdk_rx_queue *rx_queue)
+dpdk_lcore_queue_add(unsigned lcore_id, struct vr_dpdk_q_slist *q_head,
+                        struct vr_dpdk_queue *queue)
 {
-    struct vr_dpdk_lcore *lcore = vr_dpdk.lcores[lcore_id];
-    uint16_t queue_id = rx_queue - &lcore->lcore_rx_queues[0];
-
-    rte_wmb();
-
-    /* set mask to enable the queue */
-    lcore->lcore_rx_queues_mask |= 1ULL << queue_id;
-
-    /* increase the number of RX queues */
-    lcore->lcore_nb_rx_queues++;
-}
-
-/* Add a TX queue to a lcore */
-void
-dpdk_lcore_tx_queue_add(unsigned lcore_id, struct vr_dpdk_tx_queue *tx_queue)
-{
-    struct vr_dpdk_lcore *lcore = vr_dpdk.lcores[lcore_id];
-    unsigned vif_idx = tx_queue->txq_vif->vif_idx;
-    struct vr_dpdk_tx_queue *prev_tx_queue;
-    struct vr_dpdk_tx_queue *cur_tx_queue;
+    unsigned vif_idx = queue->q_vif->vif_idx;
+    struct vr_dpdk_queue *prev_queue;
+    struct vr_dpdk_queue *cur_queue;
 
     /* write barrier */
     rte_wmb();
 
     /* add queue to the list */
-    if (SLIST_EMPTY(&lcore->lcore_tx_head)) {
+    if (SLIST_EMPTY(q_head)) {
         /* insert first queue */
-        SLIST_INSERT_HEAD(&lcore->lcore_tx_head, tx_queue, txq_next);
+        SLIST_INSERT_HEAD(q_head, queue, q_next);
     } else {
         /* sort TX queues by vif_idx to optimize CPU cache usage */
-        prev_tx_queue = NULL;
-        SLIST_FOREACH(cur_tx_queue, &lcore->lcore_tx_head, txq_next) {
-            if (cur_tx_queue->txq_vif->vif_idx < vif_idx)
-                prev_tx_queue = cur_tx_queue;
+        prev_queue = NULL;
+        SLIST_FOREACH(cur_queue, q_head, q_next) {
+            if (cur_queue->q_vif->vif_idx < vif_idx)
+                prev_queue = cur_queue;
             else
                 break;
         }
         /* insert new queue */
-        if (prev_tx_queue == NULL)
-            SLIST_INSERT_HEAD(&lcore->lcore_tx_head, tx_queue, txq_next);
+        if (prev_queue == NULL)
+            SLIST_INSERT_HEAD(q_head, queue, q_next);
         else
-            SLIST_INSERT_AFTER(prev_tx_queue, tx_queue, txq_next);
+            SLIST_INSERT_AFTER(prev_queue, queue, q_next);
     }
 }
 
+/* Flush and remove TX queue from a lcore */
+static void
+dpdk_lcore_tx_queue_remove(struct vr_dpdk_lcore *lcore,
+                            struct vr_dpdk_queue *tx_queue)
+{
+    tx_queue->txq_ops.f_tx = NULL;
+    SLIST_REMOVE(&lcore->lcore_tx_head, tx_queue, vr_dpdk_queue,
+        q_next);
+    rte_wmb();
```

Comment:
Why wmb needed here - this is happening on the lcore that owns the tx queue (and no other lcore should be looking at its tx_queue list)?